.circleci/config.yml

```diff
@@ -253,6 +253,87 @@ jobs:
+    machine:
+      image: << pipeline.parameters.linux-machine-image >>
+    steps:
+      - checkout
+      - checkout-submodules
+      - npm-install
+      - run:
+          name: Build JS Bundle
+          command: npm run bundle:android
+          # This command may take a long time, so prevent CircleCI from killing
+          # the build if it doesn't receive output for more than its default
+          # ten minutes.
+          no_output_timeout: 30m
+      - run:
+          name: Get cache id
+          command: |
+            echo $CIRCLE_SHA1 > cache_id
+            cat cache_id
+      - save_cache:
+          name: Cache React Native Android Bridge JS Bundle
+          key: android-js-bundle-{{ checksum "cache_id" }}
+          paths:
+            # The build process for the react-navite-bridge requires the App.js
+            # bundle and also access to the node_modules folder.
+            - bundle/android/App.js
+            - gutenberg/node_modules/
+  android-publish-react-native-bridge:
+    # This job gets the JS bundle from the cache and uses it to assemble the
+    # react-native-bridge Android AAR to upload for binary distribution.
+    docker:
+    - image: << pipeline.parameters.android-docker-image >>
+    steps:
+      - checkout
+      - checkout-submodules
+      # Get the JS bundle from the CI cache.
+      # See `build-android-js-bundle` for details
+      - run:
+          name: Get cache id
+          command: |
+            echo $CIRCLE_SHA1 > cache_id
+            cat cache_id
+      - restore_cache:
+          name: Restore React Native Android Bridge JS Bundle from cache
+          key: android-js-bundle-{{ checksum "cache_id" }}
+      - run:
+          name: Build React Native Bridge & Upload to Bintray
+          command: |
+            cd gutenberg/packages/react-native-bridge/android
+
+            # Use a different versioning style for builds from PRs to allow
+            # developers to iterate faster.
+            #
+            # All those dev builds will be deleted once the PR is merged by
+            # https://github.com/Automattic/bintray-garbage-collector/
+            #
+            # If $VERSION is not set, that is, if we're not on a PR but a merge
+            # commit on develop or trunk, the version will be the value from
+            # the source code.
+            branch=$CIRCLE_BRANCH
+            if [[ -n "$CIRCLE_PULL_REQUEST" ]]; then
+              PR_NUMBER=$(basename $CIRCLE_PULL_REQUEST)
+              # Making the assumption that $CIRCLE_SHA1 will always be
+              # available.
+              VERSION="$PR_NUMBER-$CIRCLE_SHA1"
+            elif [[ "$branch" != "trunk" && "$branch" != "develop" ]]; then
+              # This happens on the first push of a new branch, when there
+              # isn't a PR open for it yet.
+              echo "Running on a feature branch with no open PR: skipping Bintray upload"
+              exit 0
+            fi
```

Comment:
@oguzkocer wow, I totally missed this! Thank you!

This is a port of the code from the `WordPress-Utils-Android`, where [we do everything in a single job](https://github.com/wordpress-mobile/WordPress-Utils-Android/blob/develop/.circleci/config.yml#L39-L101). I missed the possible wastefulness of running the previous step if we're not on the right branch.

> is there a way to not run the tasks at all if we are not going to publish the binary? I don't know how often that'll happen

There definitely is a way to do this. I don't know how often that'll happen either, mainly because I'm not sure whether we should build on every PR and merge to the main branch, or whether we want a manual step to trigger the publishing. @mchowning @cameronvoell can you chime in on what would be the best workflow?

I'm leaning towards always building but not making those jobs a [required status check](https://docs.github.com/en/github/administering-a-repository/managing-a-branch-protection-rule) for the PR merge. The reason I prefer this over a manual step is that it's seamless. I don't want people to have to remember to do things (I'm the first to forget tasks all the time unless I write them down) or to have to wait for a binary to be ready before being able to use it if it can be avoided.

As I discuss in https://github.com/wordpress-mobile/gutenberg-mobile/issues/2971, eventually it would be good to have some way to know whether the AAR should be built, but I don't know where to start for there and I think it fits into a dedicated PR.

Reply:
I have a working version of a setup where we only do all the steps from `npm install` if necessary.

Unfortunately, I had to use the Ubuntu machine for that, so there's a bit more overhead. I was hoping to be able to skip the checkout step as well, but I haven't figured out how. Finally, it's quite slow because it doesn't access the npm cache. I plan to work on that later on.

That's all to say, it works, but is not "right", nor fast. Yet.

Keen to hear what you think of this approach. cc @ceyhun.

https://github.com/wordpress-mobile/gutenberg-mobile/compare/mokagio/publish-to-bintray-with-script...mokagio/publish-to-bintray-with-script-demo-2

Reply:
I probably miss some context about the workflow we want to follow 😅, but I wonder if simply adding a filter to the workflow to make it run only on `develop` and `trunk` branches would work here. 
IIRC PRs _should_ be built regardless of any filters, so you shouldn't need to check this at all. 

To answer your question: you can exit any job gracefully using `circleci-agent step halt`.

Reply:
> Keen to hear what you think of this approach. cc @ceyhun.

@mokagio That looks great and seems to also have very few commands! 
I wanted to also mention that have the [android-device-checks](https://github.com/wordpress-mobile/gutenberg-mobile/blob/89a07397df85a6a4f116f81865f4cb8cfc4a446b/.circleci/config.yml#L114-L137) job where we, with a counterintuitive approach, ran a docker container inside the Ubuntu machine just to build the APK and carried out other tasks back in Ubuntu. It allowed us to overcome the memory issues and also to skip installing Android tooling. Not sure if that makes sense here, or even if it can speed things up 🙂 One piece to improve in that job could be `sudo chown -R circleci:circleci .` part to only include the folders that are really needed.

Reply:
> I probably miss some context about the workflow we want to follow 😅, but I wonder if simply adding a filter to the workflow to make it run only on develop and trunk branches would work here.

To be honest, I'm not sure what workflow we want to follow exactly either. @mchowning @cameronvoell do you mind chiming in on this? What makes more sense for the way Gutenberg mobile operate? cc @oguzkocer 

I'm also curious of how we should version the binaries built from `develop` or `trunk`. In WordPress-Utils-Android, we have a convention to always bump the version before merging into `develop` or `trunk`. Is there something similar here? Otherwise, we might want to start the AAR publishing from a tag.

I think versioning is an important but orthogonal problem. We can merge this and start using it in the real world even if the versioning workflow is not ironed out.

> IIRC PRs should be built regardless of any filters, so you shouldn't need this check at all.

Yeah, that's correct. That checks is to skip publishing packages for branches that are pushed without a PR open, for which CI starts a build anyway. The problem with publishing a package for these branches, which I estimate are either test branches or branches in the early WIP state, is that if we don't prepend the PR number to the version, then we don't have a way to automatically delete them with [the garbage collector](https://github.com/Automattic/bintray-garbage-collector/).

Reply:
@loremattei, also, thanks for this

> To answer your question: you can exit any job gracefully using `circleci-agent step halt`.

I knew there was something like that but couldn't find where we discussed about it.